File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] arr) {
     Arrays.sort(arr);
        List<List<Integer>> triplets = new ArrayList<>();
        for (int i = 0; i < arr.length - 2; i++) {
            if (i > 0 && arr[i] == arr[i - 1]) 
                continue;
            searchPair(arr, -arr[i], i + 1, triplets);
        }
        return triplets;
    }

    private void searchPair(int[] arr, int target, int left, List<List<Integer>> triplets) {
        int right = arr.length - 1;
        while (left < right) {
            int currentSum = arr[left] + arr[right];
            if (currentSum == target) { 
                triplets.add(Arrays.asList(-target, arr[left], arr[right]));
                left++;
                right--;
                while (left < right && arr[left] == arr[left - 1])
                    left++;  
                while (left < right && arr[right] == arr[right + 1])
                    right--;  
            } else if (target > currentSum)
                left++;  
            else
                right--;  
        }
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for I in range(len(nums)):
            if I > 0 and nums[I] == nums[I-1]:
                continue
            L = I + 1
            R = len(nums)-1
            while L < R:
                threeSum = nums[I] + nums[L] + nums[R]
                if threeSum > 0:
                    R -= 1
                elif threeSum < 0:
                    L += 1
                else:
                    res.append([nums[I], nums[L], nums[R]])
                    L += 1
                    while nums[L]==nums[L-1] and L < R:
                        L += 1
                    
        return res
        